storage/src/mmr/bitmap.rs

```diff
@@ -77,19 +96,98 @@ impl<H: CHasher, const N: usize> Bitmap<H, N> {
 
-    /// Return the number of bitmap chunks that have been pruned.
-    #[inline]
-    fn pruned_chunks(&self) -> usize {
-        self.mmr.pruned_to_pos() as usize
+    /// Initialize a fully pruned bitmap from the metadata in the given partition.
+    ///
+    /// The metadata must store the number of pruned chunks and the pinned hashes corresponding to
+    /// that pruning boundary.
+    pub async fn init_pruned<C: RStorage + Metrics + Clock>(
+        context: C,
+        partition: String,
+    ) -> Result<Self, Error> {
+        let metadata_cfg = MConfig { partition };
+        let metadata = Metadata::init(context.with_label("bitmap_metadata"), metadata_cfg).await?;
+
+        let key: U64 = U64::new(PRUNED_CHUNKS_PREFIX, 0);
+        let pruned_chunks = match metadata.get(&key) {
+            Some(bytes) => u64::from_be_bytes(bytes.as_ref().try_into().unwrap()),
+            None => {
+                warn!("bitmap metadata does not contain pruned chunks, initializing as empty");
+                0
+            }
+        } as usize;
+        if pruned_chunks == 0 {
+            return Ok(Self::new());
+        }
+        let mmr_size = leaf_num_to_pos(pruned_chunks as u64);
+
+        let mut pinned_nodes = Vec::new();
+        for (index, pos) in Proof::<H>::nodes_to_pin(mmr_size).enumerate() {
+            let Some(bytes) = metadata.get(&U64::new(NODE_PREFIX, index as u64)) else {
+                error!(size = mmr_size, pos, "missing pinned node");
+                return Err(Error::MissingNode(pos));
+            };
+            let digest = H::Digest::decode(bytes.as_ref());
+            let Ok(digest) = digest else {
+                error!(
+                    size = mmr_size,
+                    pos, "could not convert node bytes to digest"
+                );
+                return Err(Error::MissingNode(pos));
+            };
+            pinned_nodes.push(digest);
+        }
+
+        metadata.close().await?;
+
+        let mmr = Mmr::<H>::init(Vec::new(), mmr_size, pinned_nodes);
+
+        Ok(Self {
+            bitmap: VecDeque::from([[0u8; N]]),
+            next_bit: 0,
+            mmr,
+            pruned_chunks,
+        })
+    }
+
+    /// Write the information necessary to restore the bitmap in its fully pruned state at its last
+    /// pruning boundary. Restoring the entire bitmap state is then possible by replaying the
+    /// retained elements.
+    pub async fn write_pruned<C: RStorage + Metrics + Clock>(
+        &self,
+        context: C,
+        partition: String,
+    ) -> Result<(), Error> {
+        let metadata_cfg = MConfig { partition };
+        let mut metadata =
+            Metadata::init(context.with_label("bitmap_metadata"), metadata_cfg).await?;
+        metadata.clear();
+
+        // Write the number of pruned chunks.
+        let key = U64::new(PRUNED_CHUNKS_PREFIX, 0);
+        metadata.put(
+            key,
+            Bytes::copy_from_slice(&self.pruned_chunks.to_be_bytes()),
```

Comment:
Created an issue to clean this up: https://github.com/commonwarexyz/monorepo/issues/822